src/Microsoft.AspNet.Mvc.ViewFeatures/ViewEngines/CompositeViewEngine.cs

```diff
@@ -24,61 +25,88 @@ public CompositeViewEngine(IOptions<MvcViewOptions> optionsAccessor)
         public IReadOnlyList<IViewEngine> ViewEngines { get; }
 
         /// <inheritdoc />
-        public ViewEngineResult FindPartialView(
-            ActionContext context,
-            string partialViewName)
+        public ViewEngineResult FindView(ActionContext context, string viewName, bool isMainPage)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            if (partialViewName == null)
+            if (string.IsNullOrEmpty(viewName))
             {
-                throw new ArgumentNullException(nameof(partialViewName));
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(viewName));
             }
 
-            return FindView(context, partialViewName, partial: true);
-        }
-
-        /// <inheritdoc />
-        public ViewEngineResult FindView(
-            ActionContext context,
-            string viewName)
-        {
-            if (context == null)
+            // Do not allocate in the common cases: ViewEngines contains one entry or initial attempt is successful.
+            IEnumerable<string> searchedLocations = null;
+            List<string> searchedList = null;
+            for (var index = 0; index < ViewEngines.Count; index++)
```

Comment:
Any reason to change these loops from a `foreach` to a `for`? And why use `index` instead of good old `i`?